fdbserver/DataDistribution.actor.cpp

```diff
@@ -997,6 +1000,22 @@ struct DDTeamCollection : ReferenceCounted<DDTeamCollection> {
 		return Void();
 	}
 
+	// Check if server or machine has a valid locality based on configured replication policy
+	bool isValidLocality(Reference<IReplicationPolicy> storagePolicy, LocalityData &locality) {
+		if (!locality.isValidZoneId()) {
```

Comment:
I'm not sure whether it's possible to not have a zone ID, but if I understand correctly it is not required for single replication modes. I think it may actually not be required for the three data hall modes either, at least for storage servers.

Reply:
zoneID is used to get machineID when a TCMachine is created. 
If all processes do not set the zoneID, they will be treated as on the same machine. This will break the three data hall mode because DD always first pick a machine team and get a random server team on the machine team.

Reply:
Can't you set the zone-id explicitly to an empty string? Wouldn't that be a problem? I think by default it is set to machineID.

Though if zoneID can be invalid, I would assume one could also set an invalid machineID?

Reply:
The policy at least does not seem to indicate anything about zone IDs. Instead, it only uses the data hall value.

How about single mode?

Reply:
Single mode should be fine. 

But I still think kindly enforcing the correct setting of those localities can help avoid subtle bugs.

Reply:
> Can't you set the zone-id explicitly to an empty string? Wouldn't that be a problem? I think by default it is set to machineID.
> 
> Though if zoneID can be invalid, I would assume one could also set an invalid machineID?

As long as it is not single replica, zoneID is required. The entire team building logic is replying on the correct setting of the locality hierarchy: all processes on the same `zone` should share the same machine id. This means we cannot randomly assign UID to TCMachine (the machine concept used in team collection in DD).

Reply:
Mainly I just want to avoid adding new restrictions in a patch release. I'm not sure if it's possible to not have a zone ID, and if it isn't then this could be made into an `ASSERT` (assuming we also change the validity check to allow empty zone IDs). If it is possible, then I think this breaks clusters that used to work.

Reply:
I guess one other potential breakage in the three data hall mode is that it may work if you set zone IDs on some processes but not others. Depending on whether DD treats the missing zone ID as a zone, this could have worked before but would fail with these changes.

Reply:
I have a fix:  TCMachine will uses zoneID as its ID if zoneID is set. otherwise, it will use the SS processID as the machineID.